src/cpu.c

```diff
@@ -4,136 +4,84 @@
 #include "cpu.h"
 
 typedef void (*op_function)(void);
+
 // Function is used for instruction array initialization, not recognized by compiler
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wunused-function"
 static void UNKNOWN_OPCODE(void) {
-    fprintf(stderr, "Unallowed/Unimplemented OP Code: 0x%x\n", c.opcode);
+    fprintf(stderr, "Unallowed/Unimplemented OP Code: 0x%x\n", cpu.opcode);
 }
 #pragma GCC diagnostic pop
 
 static op_function instr_lookup[0xFF + 1] = { [0 ... 0xFF] = UNKNOWN_OPCODE };
 static op_function cb_prefixed_lookup[0xFF + 1] = { [0 ... 0xFF] = UNKNOWN_OPCODE };
 
-cpu c = {
-    0,
-    0,
-    0,
-    0,
-    0,
+Cpu cpu = {
+    .HL.word = 0,
+    .DE.word = 0,
+    .BC.word = 0,
+    .AF.word = 0,
+	0,
+	0,
     0,
     0,
-    0,
-    0,
-    0,
-    0,
-    0
 };
 
-void set_AF(uint16_t x) {
-    c.F = (uint8_t) x; // get lower octet
-    c.A = (uint8_t) (x >> 8); // get higher octet
-}
-
-uint16_t get_AF(void) {
-    return (c.A << 8) | c.F;
-}
-
-void set_BC(uint16_t x) {
-    c.C = (uint8_t) x;
-    c.B = (uint8_t) (x >> 8);
+__attribute((always_inline)) inline static void set_flag(uint8_t bit, Flag f) {
```

Comment:
What is the reasoning behind `always_inline` here? Shouldn't this be decided by the compiler (using the normal `inline` keyword as you've also done)?

Reply:
I would argue that these one-liners should always be inlined even if you are doing a non-optimized build. So basically, it makes sure that the function is inlined even when optimizations are turned off.
Here are some "interesting" results from a benchmark:
![image](https://user-images.githubusercontent.com/33261455/195954598-1c58556b-e01f-45fc-94aa-97586234a91c.png)
https://indico.cern.ch/event/386232/sessions/159923/attachments/771039/1057534/always_inline_performance.pdf

Reply:
No doubt about that, So this is basically a little bit more advanced `inline`? But isn't `inline` taken into account without the need for optimization options?

Reply:
Advanced as in, we can force the compiler in almost all cases to inline? Then yes, maybe I'm misunderstanding something here?
For the latter question: In the scenario where we compile with -O0 oder -O1 inline has no effect anymore:

> -fno-inline
>     Do not expand any functions inline apart from those marked with the always_inline attribute. This is the default when not optimizing. Single functions can be exempted from inlining by marking them with the noinline attribute.